extensions/elation/actions/createReferralOrder/createReferralOrder.ts

```diff
@@ -0,0 +1,115 @@
+> = {
+  key: 'createReferralOrder',
+  category: Category.EHR_INTEGRATIONS,
+  title: 'Create referral order',
+  description: 'Create a referral order in Elation.',
+  fields,
+  previewable: true,
+  dataPoints,
+  onActivityCreated: async (payload, onComplete, onError): Promise<void> => {    
+    try {
+      const { patient, practice, contact_name, body, authorization_for, consultant_name, specialty } = FieldsValidationSchema.parse(payload.fields)
+      const api = makeAPIClient(payload.settings)
+
+      const contactsResponse = await api.searchContactsByName({ name: contact_name })
+
+      if (contactsResponse.count === 0) {
+        throw new Error(`No contact found with the name ${contact_name}.`)
+      }
+
+      const contact = contactsResponse.results[0]
+
+      const referralOrderResponse = await api.createReferralOrder({
+        authorization_for,
+        consultant_name,
+        short_consultant_name: consultant_name,
+        practice,
+        patient,
+        specialty: { name: specialty },
+        // Setting these to null as we don't know whether they are useful or not.
+        // If needed we can add them to the action fields.
+        auth_number: null,
+        date_for_reEval: null,
+      })
+  
+      await api.postNewLetter({
+        patient,
+        practice,
+        body,
+        referral_order: referralOrderResponse.id,
+        send_to_contact: { id: contact.id },
+        letter_type: 'referral',
+      })
+
+      await onComplete({
+        data_points: {
+          id: String(referralOrderResponse.id),
+        },
+      })
+    } catch (err) {
+      if (err instanceof ZodError) {
+        const error = fromZodError(err)
+        await onError({
+          events: [
+            {
+              date: new Date().toISOString(),
+              text: { en: error.message },
+              error: {
+                category: 'WRONG_INPUT',
+                message: error.message,
+              },
+            },
+          ],
+        })
+      } else if (err instanceof AxiosError) {
+        const responseData = (err as AxiosError).response?.data
+        await onError({
+          events: [
+            {
+              date: new Date().toISOString(),
+              text: {
+                en: `${err.status ?? '(no status code)'} Error: ${err.message}`,
+              },
+              error: {
+                category: 'SERVER_ERROR',
+                message: `${err.status ?? '(no status code)'} Error data: ${JSON.stringify(responseData, null, 2)}`,
+              },
+            },
+          ],
+        })
+      } else {
+        const message = (err as Error).message
+        await onError({
+          events: [
+            {
+              date: new Date().toISOString(),
+              text: { en: message },
+              error: {
+                category: 'SERVER_ERROR',
+                message,
+              },
+            },
+          ],
+        })
+      }
+    }
+  },
```

Comment:
you actually don't have to wrap this whole thing in a try / catch, or use onError for Axios errors or Zod errors. these are caught automatically by the extension server and we do some things to support messages around errors. If you have special information you want to pass (such as transforming the response data from the axios request), then that's a good reason to keep it.

in the tests, i just do something like `const resp = action.onEvent(stuff); await expect(resp).rejects.toThrow...`

Reply:
if we can do a better job in supporting good error handling with Axios errors, i'd rather that code lived in the extension server so that all axios requests get the benefit of it

Reply:
Good to know. I copied an action that had all of this and it definitely felt unnecessary. I'll clean it up to remove all unnecessary handling logic.